narwhals/dataframe.py

```diff
@@ -2826,17 +2830,20 @@ def group_by(
             └─────────┴────────┘
             <BLANKLINE>
         """
-        from narwhals.expr import Expr
         from narwhals.group_by import LazyGroupBy
-        from narwhals.series import Series
 
-        flat_keys = flatten(keys)
-        if any(isinstance(x, (Expr, Series)) for x in flat_keys):
-            msg = (
-                "`group_by` with expression or Series keys is not (yet?) supported.\n\n"
-                "Hint: instead of `df.group_by(nw.col('a'))`, use `df.group_by('a')`."
-            )
-            raise NotImplementedError(msg)
+        flat_keys, kinds = self._flatten_and_extract(*keys)
+
+        if any(kind is ExprKind.FILTRATION for kind in kinds):
+            from narwhals.exceptions import ShapeError
+
+            msg = "series used as keys should have the same length as the DataFrame"
+            raise ShapeError(msg)
```

Comment:
That's what polars raises (when collecting) - should we post-pone it somehow?

Reply:
why post-pone? seems good to raise early here